cpp/include/raft/sparse/distance/utils.cuh

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/mr/device/allocator.hpp>
+#include <raft/mr/device/buffer.hpp>
+
+#include <cub/cub.cuh>
+
+#include <rmm/device_vector.hpp>
+
+namespace raft {
+namespace sparse {
+namespace distance {
+
+/**
+ * @tparam value_idx
+ * @param out
+ * @param in
+ * @param n
+ */
+template <typename value_idx, bool count_greater_than>
+__global__ void max_kernel(value_idx *out, const value_idx *in,
+                           value_idx *out_count, value_idx n,
+                           value_idx greater_than) {
+  int tid = blockDim.x * blockIdx.x + threadIdx.x;
+
+  typedef cub::BlockReduce<value_idx, 256> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  value_idx v;
+  if (tid < n) {
+    v = in[tid] - in[tid - 1];
+  } else {
+    v = 0;
+  }
+  value_idx agg = BlockReduce(temp_storage).Reduce(v, cub::Max());
+
+  if (count_greater_than) {
+    bool predicate = tid < n && v > greater_than;
+    value_idx count = __syncthreads_count(predicate);
+
+    if (threadIdx.x == 0) atomicAdd(out_count, count);
+  }
+
+  if (threadIdx.x == 0) atomicMax(out, agg);
+}
+
+template <typename value_idx, bool count_greater_than>
+inline std::pair<value_idx, value_idx> max_degree(
```

Comment:
Can we do a brief doxygen for this?

Reply:
I don't even think I'm using this as a pair of returned values anymore. I'll revert it to minimal function and add doxygen